safekeeper/src/wal_backup.rs

```diff
@@ -0,0 +1,267 @@
+use std::path::PathBuf;
+
+use postgres_ffi::xlog_utils::{XLogSegNo, PG_TLI, XLogFileName, XLogSegNoOffsetToRecPtr};
+use tokio::fs::File;
+use tokio::runtime::{Builder, Runtime};
+
+use anyhow::{Result, ensure, bail, Context};
+
+use tokio::sync::watch::{Receiver, Sender};
+use utils::{
+    lsn::Lsn,
+    zid::{ZTenantTimelineId},
+};
+
+use tracing::*;
+
+use crate::config::{RemoteStorageConfig, RemoteStorageKind};
+use crate::remote_storage::{NoopStorage, RemoteStorage, LocalFs, S3Bucket};
+use crate::{broker, SafeKeeperConf, wal_storage};
+
+
+use once_cell::sync::OnceCell;
+
+static BACKUP_RUNTIME: OnceCell<Runtime> = OnceCell::new();
+const DEFAULT_BACKUP_RUNTIME_SIZE: u32 = 16;
+
+const MIN_WAL_SEGMENT_SIZE: usize = 1 << 20; // 1MB
+const MAX_WAL_SEGMENT_SIZE: usize = 1 << 30; // 1GB
+const DEFAULT_WAL_SEGMENT_SIZE: usize = 16 << 20; // 16MB
+
+const BACKUP_ELECTION_PATH: &str = "WAL_BACKUP";
+
+async fn detect_task(
+    conf: SafeKeeperConf,
+    timeline_id: ZTenantTimelineId,
+    backup_start: Lsn,
+    mut wal_seg_size: Receiver<u32>,
+    lease_id : Option<i64>, 
+    mut lsn_durable: Receiver<Lsn>, 
+    lsn_backed_up: Sender<Lsn>,
+    ) -> Result<()> {
+
+    // TODO: antons come up with better naming
+    wal_seg_size.changed().await.expect("Failed to recieve wal segment size");
+    let wal_seg_size = *wal_seg_size.borrow();
+    let mut wal_seg_size = usize::try_from(wal_seg_size).with_context(|| format!("Couldn't convert {} to usize", wal_seg_size))?;
```

Comment:
> ?

There are a lot of such places in the loop, that just abruptly return `Err(anyhow::Error)`, stopping the loop.
We don't check for that and never restart the loop: we drop the handle of this task spawned. Neither we check if the task spawned had panicked.

IMO, we should log such cases and restart the backup loop for every timeline. Some exponential backoff could be used to avoid odd network issues.

Reply:
please check if the latest looks any better